File: src/lib/services/deployment-service.ts
import fs from 'fs/promises';
import path from 'node:path';
import { BASE_PATH } from '$lib/services/paths-service';
import type { Deployment } from '$lib/types/deployment.type';
import { nanoid } from 'nanoid';
import { getDeploymentsFromDb, getDeploymentFromDb, saveDeploymentToDb, updateDeploymentInDb, deleteDeploymentFromDb, getDeploymentByTaskIdFromDb } from './database/deployment-db-service';

const DEPLOYMENTS_DIR = path.join(BASE_PATH, 'deployments');

// Ensure directory exists
await fs.mkdir(DEPLOYMENTS_DIR, { recursive: true });

export async function createDeployment(deployment: Omit<Deployment, 'id' | 'createdAt'>): Promise<Deployment> {
	const newDeployment: Deployment = {
		...deployment,
		id: nanoid(),
		createdAt: new Date().toISOString(),
		updatedAt: new Date().toISOString()
	};

	return await saveDeploymentToDb(newDeployment);
}

export async function updateDeployment(deploymentId: string, updates: Partial<Deployment>): Promise<Deployment | null> {
	try {
		const existingDeployment = await getDeploymentFromDb(deploymentId);

		if (!existingDeployment) {
			console.error('Deployment not found:', deploymentId);
			return null;
		}

		const updatedDeployment = {
			...existingDeployment,
			...updates,
			updatedAt: new Date().toISOString()
		};

		return await saveDeploymentToDb(updatedDeployment);
	} catch (error) {
		console.error('Error updating deployment:', error);
		return null;
	}
}

export async function getDeployment(deploymentId: string): Promise<Deployment | null> {
	return await getDeploymentFromDb(deploymentId);
}

export async function getDeployments(agentId?: string): Promise<Deployment[]> {
	return await getDeploymentsFromDb(agentId);
}

export async function deleteDeployment(deploymentId: string): Promise<boolean> {
	return await deleteDeploymentFromDb(deploymentId);
}

// Helper functions for creating specific deployment types
export async function createStackDeployment(agentId: string, stackName: string, composeContent: string, envContent?: string, taskId?: string): Promise<Deployment> {
	return createDeployment({
		name: stackName,
		type: 'stack',
		status: 'pending',
		agentId,
		metadata: {
			stackName,
			composeContent,
			envContent
		},
		taskId
	});
}

export async function createContainerDeployment(agentId: string, containerName: string, imageName: string, ports?: string[], volumes?: string[], taskId?: string): Promise<Deployment> {
	return createDeployment({
		name: containerName || imageName,
		type: 'container',
		status: 'pending',
		agentId,
		metadata: {
			containerName,
			imageName,
			ports,
			volumes
		},
		taskId
	});
}

export async function createImageDeployment(agentId: string, imageName: string, taskId?: string): Promise<Deployment> {
	return createDeployment({
		name: imageName,
		type: 'image',
		status: 'pending',
		agentId,
		metadata: {
			imageName
		},
		taskId
	});
}

// Update deployment status based on task completion
export async function updateDeploymentFromTask(taskId: string, status: string, result?: any, error?: string): Promise<void> {
	try {
		// Find deployment linked to this task
		const deployment = await getDeploymentByTaskIdFromDb(taskId);

		if (!deployment) {
			// No deployment linked to this task, nothing to update
			return;
		}

		// Map task status to deployment status
		let deploymentStatus: 'pending' | 'running' | 'stopped' | 'failed' | 'completed';

		switch (status) {
			case 'running':
				deploymentStatus = 'running';
				break;
			case 'completed':
				deploymentStatus = 'completed';
				break;
			case 'failed':
				deploymentStatus = 'failed';
				break;
			default:
				deploymentStatus = 'pending';
		}

		// Update deployment in database
		await updateDeploymentInDb(deployment.id, {
			status: deploymentStatus,
			error: error || undefined
		});

		console.log(`Deployment ${deployment.id} updated to status: ${deploymentStatus}`);
	} catch (error) {
		console.error('Failed to update deployment from task:', error);
		// Don't throw as this shouldn't break the task update
	}
}

File: backend/internal/services/deployment_service.go
package services

import (
	"context"
	"errors"
	"fmt"
	"time"

	"github.com/google/uuid"
	"github.com/ofkm/arcane-backend/internal/database"
	"github.com/ofkm/arcane-backend/internal/models"
	"gorm.io/gorm"
)

type DeploymentService struct {
	db *database.DB
}

func NewDeploymentService(db *database.DB) *DeploymentService {
	return &DeploymentService{db: db}
}

func (s *DeploymentService) CreateDeployment(ctx context.Context, deployment *models.Deployment) (*models.Deployment, error) {
	deployment.ID = uuid.New().String()
	deployment.BaseModel = models.BaseModel{CreatedAt: time.Now()}

	if err := s.db.WithContext(ctx).Create(deployment).Error; err != nil {
		return nil, fmt.Errorf("failed to create deployment: %w", err)
	}

	return deployment, nil
}

func (s *DeploymentService) GetDeploymentByID(ctx context.Context, id string) (*models.Deployment, error) {
	var deployment models.Deployment
	if err := s.db.WithContext(ctx).Preload("Agent").Preload("Task").Where("id = ?", id).First(&deployment).Error; err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			return nil, fmt.Errorf("deployment not found")
		}
		return nil, fmt.Errorf("failed to get deployment: %w", err)
	}
	return &deployment, nil
}

func (s *DeploymentService) ListDeployments(ctx context.Context, agentID *string) ([]*models.Deployment, error) {
	query := s.db.WithContext(ctx).Preload("Agent").Preload("Task").Order("created_at DESC")

	if agentID != nil {
		query = query.Where("agent_id = ?", *agentID)
	}

	var deployments []*models.Deployment
	if err := query.Find(&deployments).Error; err != nil {
		return nil, fmt.Errorf("failed to list deployments: %w", err)
	}
	return deployments, nil
}

func (s *DeploymentService) UpdateDeploymentStatus(ctx context.Context, id string, status models.DeploymentStatus, error *string) error {
	updates := map[string]interface{}{
		"status":     status,
		"updated_at": time.Now(),
	}

	if error != nil {
		updates["error"] = *error
	}

	if err := s.db.WithContext(ctx).Model(&models.Deployment{}).Where("id = ?", id).Updates(updates).Error; err != nil {
		return fmt.Errorf("failed to update deployment status: %w", err)
	}

	return nil
}

func (s *DeploymentService) DeleteDeployment(ctx context.Context, id string) error {
	if err := s.db.WithContext(ctx).Delete(&models.Deployment{}, "id = ?", id).Error; err != nil {
		return fmt.Errorf("failed to delete deployment: %w", err)
	}
	return nil
}

// Helper methods for creating specific deployment types
func (s *DeploymentService) CreateStackDeployment(ctx context.Context, agentID, stackName, composeContent string, envContent *string, taskID *string) (*models.Deployment, error) {
	metadata := models.JSON{
		"stackName":      stackName,
		"composeContent": composeContent,
	}

	if envContent != nil {
		metadata["envContent"] = *envContent
	}

	deployment := &models.Deployment{
		Name:     stackName,
		Type:     models.DeploymentTypeStack,
		Status:   models.DeploymentStatusPending,
		AgentID:  agentID,
		TaskID:   taskID,
		Metadata: metadata,
	}

	return s.CreateDeployment(ctx, deployment)
}

func (s *DeploymentService) CreateContainerDeployment(ctx context.Context, agentID, containerName, imageName string, ports, volumes []string, taskID *string) (*models.Deployment, error) {
	metadata := models.JSON{
		"containerName": containerName,
		"imageName":     imageName,
		"ports":         ports,
		"volumes":       volumes,
	}

	deployment := &models.Deployment{
		Name:     containerName,
		Type:     models.DeploymentTypeContainer,
		Status:   models.DeploymentStatusPending,
		AgentID:  agentID,
		TaskID:   taskID,
		Metadata: metadata,
	}

	return s.CreateDeployment(ctx, deployment)
}

func (s *DeploymentService) CreateImageDeployment(ctx context.Context, agentID, imageName string, taskID *string) (*models.Deployment, error) {
	metadata := models.JSON{
		"imageName": imageName,
	}

	deployment := &models.Deployment{
		Name:     imageName,
		Type:     models.DeploymentTypeImage,
		Status:   models.DeploymentStatusPending,
		AgentID:  agentID,
		TaskID:   taskID,
		Metadata: metadata,
	}

	return s.CreateDeployment(ctx, deployment)
}
